libass/ass_utils.h

```diff
@@ -51,6 +51,14 @@
 
 #define FEATURE_MASK(feat) (((uint32_t) 1) << (feat))
 
+#if defined(__GNUC__)
+    #define ASSUME(x) if (!(x)) __builtin_unreachable()
```

Comment:
`__builtin_unreachable` was added in [gcc@468059b](https://gcc.gnu.org/git/?p=gcc.git&a=commit;h=468059bcbb8efdb31fd90369dc2ed122126413fe) and seems to have been first released with 4.5.0 and for clang in [llvm@bf20638966](https://github.com/llvm/llvm-project/commit/bf20638966e205e6dc18451275e5c51264f255e6) released with llvm 2.7.0. For ideal compatibility we’d need to use the following somewhat unwieldy check:

```C
#if (defined(__GNUC__) && ((__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5))) ||
     (defined(__clang__) && __has_builtin(__builtin_unreachable))
```

While clang [got `__has_builtin`](https://github.com/llvm/llvm-project/commit/b6f77af5327f411aa07b81b25b85b016113dace5) before `__builtin_unreachable`, gcc unfortunately only [gained it a decade later](https://gcc.gnu.org/git/?p=gcc.git&a=commit;h=ad1539d5555a161cf6851de8995641d6dfe792d9).

Given this is about optimisation though and even gcc supports `__has_builtin` for around 4 years, stubbing out `__has_builtin` for older compilers and using that for `defined(__GNUC__) || defined(__clang__)` might be fine too

Reply:
Done (using `__has_builtin`).